---
title: "Quality Control and Filtering"
output: github_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Dependencies

Load the major packages needed for this analysis first. Additional packages will be loaded as needed.

```{r, message=FALSE, warning=FALSE, echo = TRUE}
# dependencies needed
library(SingleCellExperiment)
library(scater)
library(scran)
library(uwot)
library(Rtsne)
```

## Importing Mouse Genome Annotation
```{r, echo=TRUE, message = FALSE}
library(AnnotationHub) # load the ensembldb
ah = AnnotationHub() # create annotation hub object


```


```{r, echo=TRUE}

#search for and then load emsembl db v98 for mouse
dbNumber<-names(query(ah, c('Ensembl', 'musculus', 'EnsDb', '98')))
eDb<-ah[[dbNumber]] 
keytypes(eDb) # remind me what keys are availble to select records with
columns(eDb)
```

```{r}
unfiltered <- readRDS(file="./processed_data/guiu_unfiltered.rds")
```


```{r, echo=TRUE}
## Goal - to retrieve the chromosome location of each gene in sce.guiu to see if it is mtDNA
chrom<-mapIds(eDb, keys = rowData(unfiltered)$ID, column = "SEQNAME", keytype = "GENEID")
# genes belonging to mtDNA as determined from chrom/edb
mito<-which(chrom=="MT")
```

```{r, echo=TRUE}
# Compute the QC metrics per cell
library(scater)
metrics <- perCellQCMetrics(unfiltered, subsets=list(Mito=mito))
head(metrics)


metrics.lib <- isOutlier(metrics$sum, log=TRUE, nmads = 2, type="lower")
attr(metrics.lib, "thresholds")
metrics.exprs <- isOutlier(metrics$detected, log=TRUE, nmads = 2,type="lower")
attr(metrics.exprs, "thresholds")
metrics.mito <- isOutlier(metrics$subsets_Mito_percent,nmads = 2, type="higher")
attr(metrics.mito, "thresholds")


discard <- metrics.lib | metrics.exprs | metrics.mito
```

```{r, echo=TRUE}
# Summarize the number of cells removed by which reason.
DataFrame(libSize=sum(metrics.lib), exprs=sum(metrics.exprs),
           mitoPercent=sum(metrics.mito), total=sum(discard))
```

```{r, echo=TRUE}
# column bind the states from perCellQC to the original sce now called unfiltered
colData(unfiltered) <- cbind(colData(unfiltered), metrics)
# add a new column called discard that includes the logical for metrics.mito
unfiltered$discard <- discard
```

```{r, echo=TRUE}
# take a look at what the filtering did with plots colored by discard identity
gridExtra::grid.arrange(
  plotColData(unfiltered, y="sum", colour_by="discard") +
    scale_y_log10() + ggtitle("total count"),
  plotColData(unfiltered, y="detected", colour_by="discard") +
    scale_y_log10() + ggtitle("detected features"),
  plotColData(unfiltered, y="subsets_Mito_percent",
    colour_by="discard") + ggtitle("percent mito"),
  ncol=2
)
```

```{r, echo=TRUE}
#compare filtering metrics to each other
plotColData(unfiltered, x="sum", y="detected",
            colour_by="discard") + scale_x_log10()
plotColData(unfiltered, x="sum", y="subsets_Mito_percent",
            colour_by="discard") + scale_x_log10()

```

```{r, echo=TRUE}
# Keeping the columns we DON'T want to discard.
filtered <- unfiltered[,!discard]
dim(unfiltered) # more cells
dim(filtered) # less cells!
```

```{r, echo=TRUE}

saveRDS(filtered, file="./processed_data/guiu_filtered.rds") # save the sce post QC
```


```{r}
sessionInfo()
```
